### ISSUE TITLE
progressDeadlineSeconds not respected, failed deployment not detected

### ISSUE THREAD

**Issue by brunowoo** [kubernetes/kubernetes#40207]
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug report

**Kubernetes version** (use `kubectl version`): both client and server on v1.5.2

**Environment**:
- **Cloud provider or hardware configuration**: AWS
- **OS** (e.g. from /etc/os-release): Ubuntu 16.04LTS
- **Kernel** (e.g. `uname -a`):  4.4.0-53-generic #74-Ubuntu SMP Fri Dec 2 15:59:10 UTC 2016 x86_64
- **Install tools**: kops 1.5.0-alpha3

**What happened**:
Added the following field to my deployment spec: `minReadySeconds: 5` and `progressDeadlineSeconds: 120`, with a pod spec for a single-container pod with a non existing-image:tag (the image exists in my private dockerhub registry, the tag does not). I then deployed this update using `kubectl apply -f <filename>.yaml`  
The pods fail (first with ErrImagePull and then with ImagePullBackOff):
```
NAME                          READY     STATUS         RESTARTS   AGE  
shotgun-2497328708-bvq79      0/1       ImagePullBackOff   0          10s 
shotgun-2923704814-8q88v      1/1       Running        0          19h
```
The replicaset just keeps waiting:
```
Name:		shotgun-2497328708
Namespace:	test
Image(s):	shotgun:v99
Selector:	app=shotgun,pod-template-hash=2497328708
Labels:		app=shotgun
		pod-template-hash=2497328708
		version=v99
Replicas:	1 current / 1 desired
Pods Status:	0 Running / 1 Waiting / 0 Succeeded / 0 Failed
No volumes.
Events:
  FirstSeen	LastSeen	Count	From				SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  10m		10m		1	{replicaset-controller }			Normal		SuccessfulCreate	Created pod: shotgun-2497328708-bvq79
```
And the deployment reports it is progressing, far beyond the deadline set in the spec:
```
Name:			shotgun
Namespace:		test
CreationTimestamp:	Tue, 03 Jan 2017 11:40:10 +0100
Labels:			app=shotgun
Selector:		app=shotgun
Replicas:		1 updated | 1 total | 1 available | 1 unavailable
StrategyType:		RollingUpdate
MinReadySeconds:	5
RollingUpdateStrategy:	0 max unavailable, 1 max surge
Conditions:
  Type		Status	Reason
  ----		------	------
  Available 	True	MinimumReplicasAvailable
  Progressing 	True	NewReplicaSetAvailable
OldReplicaSets:	shotgun-2923704814 (1/1 replicas created)
NewReplicaSet:	shotgun-2497328708 (1/1 replicas created)
Events:
  FirstSeen	LastSeen	Count	From				SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----				-------------	--------	------			-------
  17h		11m		2	{deployment-controller }			Normal		ScalingReplicaSet	Scaled up replica set shotgun-2497328708 to 1
```
The status from `kubectl get deployment/shotgun --namespace=test -o yaml` reports the following:
```
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: 2017-01-20T09:29:43Z
    lastUpdateTime: 2017-01-20T09:29:43Z
    message: Replica set "shotgun-2497328708" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
```
This all stays the same, even 20min later.

**What you expected to happen**:
The pods to report as failed back to the replicaset instead of waiting and the deployment to mark it is not progressing after seeing no progress after 120 seconds.

**How to reproduce it** (as minimally and precisely as possible):
Pushing an update to a deployment with a non-existing image (possible private?) and waiting for the progressDeadlineSeconds to expire.

**Comment by brunowoo** [comment]
@kargakis thanks for the quick response. Is it as intended that the pods don't get listed as failed in the ReplicaSet?

**Comment by kargakis** [comment]
ReplicaSets surface a condition when they can't create or delete a pod.
Eventually I would like to surface error conditions such as
ImagePullBackOff or CrashLoopBackOff as well but the ReplicaSets can't make
any decisions based on those conditions. Deployments, the resource
responsible for making sure that ReplicaSets are successfully deployed, are
more likely to act based on a failure condition. Right now though we just
expose the failure conditions and a progressDeadlineExceeeded conditon if
progressDeadlineSeconds is set. In the future, Deployments will be able to
automatically rollback ReplicaSets when a perma-failed condition or a
progressDeadlineExceeded condition is met.

**Comment by kargakis** [comment]
Cherry-picking back to 1.5 in https://github.com/kubernetes/kubernetes/pull/40625 - should be included in 1.5.3
